JoinColony/colonyCDapp
# Feat: Implement edit locked expenditure motion saga

## Description

[Related block-ingestor PR](https://github.com/JoinColony/block-ingestor/pull/189)

Changes to expenditures after locking can be made using `setExpenditureState` contracts function. This PR implements an `editLockedExpenditureMotion` saga that accepts an array of updated payouts and creates a motion which, upon finalizing, calls the contracts function and updates the expenditure accordingly.

## Testing

1. Install the reputation weighting extension
2. Create a new expenditure (The expenditure ID will be logged to the console)
3. Enter the expenditure ID and lock the expenditure.
4. Run this query to verify the expenditure has been created with no motions attached, and one payout.

```
query MyQuery {
  listExpenditures(filter: {nativeId: {ge: 33}}) {
    items {
      nativeId
      motions {
        items {
          motionStateHistory {
            hasPassed
          }
          editedExpenditureSlots {
            claimDelay
            id
            payoutModifier
            recipientAddress
            payouts {
              amount
              isClaimed
              networkFee
              tokenAddress
            }
          }
        }
      }
      slots {
        payouts {
          amount
          isClaimed
          networkFee
          tokenAddress
        }
        claimDelay
        recipientAddress
        id
        payoutModifier
      }
      status
    }
  }
}

```

5. Click the 'Edit via motion' button and follow the link logged in the console
6. Run the query again to verify that the 'editedExpenditureSlots' field of the attached motion is correctly storing all of the proposed changes
7. Stake both sides and vote to support the motion. Then finalize the motion.
8. Run the query a third time to verify that the motion has passed, and the payouts have updated.

## Diffs

**New stuff** ✨

* EditLockedExpenditureMotion saga

Resolves #1976
